src/recordlinker/linkage/models.py

```diff
@@ -47,3 +44,48 @@ class BlockingValue(Base):
     patient_id: orm.Mapped[int] = orm.mapped_column(ForeignKey("mpi_patient.id"))
     blockingkey_id: orm.Mapped[int] = orm.mapped_column(ForeignKey("mpi_blocking_key.id"))
     value: orm.Mapped[str] = orm.mapped_column(String(50), index=True)
+
+class Algorithm(Base):
+    __tablename__ = "algorithm"
+
+    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
+    is_default: orm.Mapped[bool] = orm.mapped_column(default=False)
+    label: orm.Mapped[str] = orm.mapped_column(String(255), unique=True)
+    description: orm.Mapped[str]
+
+def check_only_one_default(connection, target):
+    """
+    Check if there is already a default algorithm before inserting or updating.
+
+    Called before an insert or update operation on the
+    Algorithm table. If the `is_default` attribute of the target object is
+    set to True, it checks the database to ensure that no other algorithm
+    is marked as default. If another default algorithm exists, an exception
+    is raised to prevent the operation.
+
+    Parameters:
+    connection: The database connection being used for the operation.
+    target: The instance of the Algorithm class being inserted or updated.
+    
+    Raises:
+    Exception: If another algorithm is already marked as default.
+    """
+     
+    if target.is_default:
+        existing_default = connection.execute("SELECT COUNT(*) FROM algorithm WHERE is_default = TRUE").scalar()
+        if(existing_default > 0):
```

Comment:
This will fail to catch the "before_update" use case.  If the user is updating an Algorithm, that is already set as the default.  If the update say, changes the description on the algorithm, but leave the `is_default=True` as is.  This if block will pass and the update will fail.

Reply:
updated